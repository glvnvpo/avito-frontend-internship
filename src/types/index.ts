// @flow

export type ID = number | string;

export type URL = string;

export type Story = {
    id: ID;
    title: string;
    by: string;
    time: number | string;
    score: number | string;
    kids?: Array<number | string>;
    text?: string;
    url?: string;
    descendants?: number | string;
}

export type ChildComment = {
    id: number | string;
    by: string;
    text: string;
    time: number | string;
    kids?: Array<number | string>;
}

export type Comment = {
    id: number | string;
    by: string;
    text: string;
    time: number | string;
    kids?: Array<number | string>;
    children?: Array<ChildComment>;
    showChildComment?: boolean;
    isLoadingChildren?: boolean;
}