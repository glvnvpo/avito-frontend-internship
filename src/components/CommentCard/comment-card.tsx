// @flow

import React, {FC, ReactNode} from 'react';
import {Button, Comment} from 'antd';
import {isEmpty} from 'lodash';
import parse from 'html-react-parser';
import './styles.scss';
import {Colors, Sizes, Spinner} from '../Spinner';
import {Comment as CommentType} from '../../types';
import {getDateFromTimestamp} from '../../helpers/get-date-from-timestamp';

type Props = {
	comment: CommentType;
	isParent?: boolean;
	showAnswers?: (parentComment: CommentType) => void;
	children?: ReactNode;
}

export const CommentCard: FC<Props> = ({comment, isParent=true, showAnswers, children, ...rest}) => {

	let {by, time, kids, deleted, dead, isLoadingChildren, showChildComment} = comment;

	const getBtnText = (isOpen: boolean | undefined):string => {
		if (isOpen) {
			return 'Hide answers';
		}
		else return 'Show answers';
	};

	const actions = (isParent && !isEmpty(kids) && showAnswers) ? [
		<Button onClick={() => showAnswers(comment)}>
			{ isLoadingChildren ? <Spinner size={Sizes.SMALL} color={Colors.BLUE}/> : getBtnText(showChildComment) }
		</Button>
	] : undefined;
	
	const getContent = (comment: CommentType) => {
		const {text, deleted, dead} = comment;

		if (deleted) {
			return 'This comment was deleted';
		}

		if (dead) {
			return 'Comment not available';
		}

		return parse(text);
	};

	return (
		<Comment
			className={`comment-card ${deleted && 'deleted'} ${dead && 'dead'}`}
			author={by}
			content={getContent(comment)}
			datetime={getDateFromTimestamp(time)}
			actions={actions}
			{...rest}
		>
			{children}
		</Comment>
	);
};