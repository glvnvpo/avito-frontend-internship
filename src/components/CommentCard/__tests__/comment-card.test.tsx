import React from 'react';
import {mount} from 'enzyme';
import {CommentCard} from '../comment-card';
import {Comment} from '../../../types';

describe('CommentCard', () => {
	let wrapper;

	const comment = {
		id: 123,
		by: 'coolUser',
		text: 'someText',
		time: 1666689031,
		kids: [124],
		showChildComment: true
	};
	const childComment = {
		id: 124,
		by: 'coolUser124',
		text: 'someText124',
		time: 1666689031,
		parent: 123
	};
	const showAnswers = jest.fn();
	
	const getComponent = (parentComment: Comment) => {
		return mount(
			<CommentCard comment={parentComment} showAnswers={showAnswers}>
				<CommentCard comment={childComment} isParent={false}/>
			</CommentCard>);
	};
	
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it('should render correct comment tree (opened)', () => {
		wrapper = getComponent(comment);
		expect(wrapper).toMatchSnapshot();
	});

	it('should render correct button text when comment tree is closed', () => {
		const parentComment = {
			...comment,
			showChildComment: false
		};
		wrapper = getComponent(parentComment);
		expect(wrapper.find('.ant-btn').text()).toBe('Show answers');
	});
	
	it('should call showAnswers with parent comment', () => {
		wrapper = getComponent(comment);
		wrapper.find('.ant-btn').simulate('click');
		expect(showAnswers).toHaveBeenCalledWith(comment);
	});

	it('should render spinner when loading children comments', () => {
		const parentComment = {
			...comment,
			showChildComment: false,
			isLoadingChildren: true
		};
		wrapper = getComponent(parentComment);
		expect(wrapper.find('Spinner')).toHaveLength(1);
	});

	it('should return correct text when comment is deleted', () => {
		const parentComment = {
			...comment,
			deleted: true
		};
		wrapper = getComponent(parentComment);
		expect(wrapper.find('Comment').at(0).prop('content')).toBe('This comment was deleted');
	});

	it('should return correct text when comment is dead', () => {
		const parentComment = {
			...comment,
			dead: true
		};
		wrapper = getComponent(parentComment);
		expect(wrapper.find('Comment').at(0).prop('content')).toBe('Comment not available');
	});
});
