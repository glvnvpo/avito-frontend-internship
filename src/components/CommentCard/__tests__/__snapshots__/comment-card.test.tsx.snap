// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentCard should render correct comment tree (opened) 1`] = `
<CommentCard
  comment={
    {
      "by": "coolUser",
      "id": 123,
      "kids": [
        124,
      ],
      "showChildComment": true,
      "text": "someText",
      "time": 1666689031,
    }
  }
  showAnswers={[MockFunction]}
>
  <Comment
    actions={
      [
        <Button
          onClick={[Function]}
        >
          Hide answers
        </Button>,
      ]
    }
    author="coolUser"
    className="comment-card undefined undefined"
    content="someText"
    datetime="25.10.2022"
  >
    <div
      className="ant-comment comment-card undefined undefined"
    >
      <div
        className="ant-comment-inner"
      >
        <div
          className="ant-comment-content"
        >
          <div
            className="ant-comment-content-author"
          >
            <span
              className="ant-comment-content-author-name"
            >
              coolUser
            </span>
            <span
              className="ant-comment-content-author-time"
            >
              25.10.2022
            </span>
          </div>
          <div
            className="ant-comment-content-detail"
          >
            someText
          </div>
          <ul
            className="ant-comment-actions"
          >
            <li
              key="action-0"
            >
              <Button
                onClick={[Function]}
              >
                <ForwardRef
                  disabled={false}
                >
                  <InternalWave
                    disabled={false}
                  >
                    <button
                      className="ant-btn ant-btn-default"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      <LoadingIcon
                        existIcon={false}
                        loading={false}
                        prefixCls="ant-btn"
                      >
                        <CSSMotion
                          motionName="ant-btn-loading-icon-motion"
                          onAppearActive={[Function]}
                          onAppearStart={[Function]}
                          onEnterActive={[Function]}
                          onEnterStart={[Function]}
                          onLeaveActive={[Function]}
                          onLeaveStart={[Function]}
                          removeOnLeave={true}
                          visible={false}
                        >
                          <DomWrapper />
                        </CSSMotion>
                      </LoadingIcon>
                      <span
                        key=".0"
                      >
                        Hide answers
                      </span>
                    </button>
                  </InternalWave>
                </ForwardRef>
              </Button>
            </li>
          </ul>
        </div>
      </div>
      <div
        className="ant-comment-nested"
      >
        <CommentCard
          comment={
            {
              "by": "coolUser124",
              "id": 124,
              "parent": 123,
              "text": "someText124",
              "time": 1666689031,
            }
          }
          isParent={false}
        >
          <Comment
            author="coolUser124"
            className="comment-card undefined undefined"
            content="someText124"
            datetime="25.10.2022"
          >
            <div
              className="ant-comment comment-card undefined undefined"
            >
              <div
                className="ant-comment-inner"
              >
                <div
                  className="ant-comment-content"
                >
                  <div
                    className="ant-comment-content-author"
                  >
                    <span
                      className="ant-comment-content-author-name"
                    >
                      coolUser124
                    </span>
                    <span
                      className="ant-comment-content-author-time"
                    >
                      25.10.2022
                    </span>
                  </div>
                  <div
                    className="ant-comment-content-detail"
                  >
                    someText124
                  </div>
                </div>
              </div>
            </div>
          </Comment>
        </CommentCard>
      </div>
    </div>
  </Comment>
</CommentCard>
`;
